File: ts/packages/nostr-pubsub/src/invwant.ts
import type { SourceId } from './types.js';

export const DEFAULT_INV_WANT_HOP_LIMIT = 16;

export interface PubsubContentKey {
  streamId: string;
  origin: SourceId;
  seq: number;
}

export interface PubsubInventory {
  key: PubsubContentKey;
  payloadBytes: number;
  hopLimit: number;
}

export interface PubsubWant {
  key: PubsubContentKey;
}

export interface PubsubFrame {
  key: PubsubContentKey;
  payload: Uint8Array;
  hopLimit: number;
}

export type InvWantMessage =
  | { type: 'inventory'; inventory: PubsubInventory }
  | { type: 'want'; want: PubsubWant }
  | { type: 'frame'; frame: PubsubFrame };

export function createContentKey(streamId: string, origin: SourceId, seq: number): PubsubContentKey {
  return { streamId, origin, seq };
}

export function createInventory(
  key: PubsubContentKey,
  payloadBytes: number,
  hopLimit: number,
): PubsubInventory {
  return { key: cloneContentKey(key), payloadBytes, hopLimit };
}

export function wantFromInventory(inventory: PubsubInventory): PubsubWant {
  return { key: cloneContentKey(inventory.key) };
}

export function createWant(key: PubsubContentKey): PubsubWant {
  return { key: cloneContentKey(key) };
}

export function createFrame(
  key: PubsubContentKey,
  payload: Uint8Array | ArrayLike<number>,
  hopLimit: number,
): PubsubFrame {
  return {
    key: cloneContentKey(key),
    payload: payload instanceof Uint8Array ? new Uint8Array(payload) : Uint8Array.from(payload),
    hopLimit,
  };
}

export function inventoryFromFrame(frame: PubsubFrame): PubsubInventory {
  return createInventory(frame.key, frame.payload.byteLength, frame.hopLimit);
}

export function invWantMessageKey(message: InvWantMessage): PubsubContentKey {
  switch (message.type) {
    case 'inventory':
      return message.inventory.key;
    case 'want':
      return message.want.key;
    case 'frame':
      return message.frame.key;
  }
}

export function invWantMessageStreamId(message: InvWantMessage): string {
  return invWantMessageKey(message).streamId;
}

function cloneContentKey(key: PubsubContentKey): PubsubContentKey {
  return { streamId: key.streamId, origin: key.origin, seq: key.seq };
}

File: crates/nostr-pubsub/src/inv_want.rs
use async_trait::async_trait;

use crate::{EventSource, PublishReport, Result, SourceId};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PubsubStreamId(pub String);

impl PubsubStreamId {
    #[must_use]
    pub fn new(stream_id: impl Into<String>) -> Self {
        Self(stream_id.into())
    }

    #[must_use]
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PubsubContentKey {
    pub stream_id: PubsubStreamId,
    pub origin: SourceId,
    pub seq: u64,
}

impl PubsubContentKey {
    #[must_use]
    pub fn new(stream_id: impl Into<String>, origin: impl Into<String>, seq: u64) -> Self {
        Self {
            stream_id: PubsubStreamId::new(stream_id),
            origin: SourceId::new(origin),
            seq,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PubsubInventory {
    pub key: PubsubContentKey,
    pub payload_bytes: u64,
    pub hop_limit: u8,
}

impl PubsubInventory {
    #[must_use]
    pub fn new(key: PubsubContentKey, payload_bytes: u64, hop_limit: u8) -> Self {
        Self {
            key,
            payload_bytes,
            hop_limit,
        }
    }

    #[must_use]
    pub fn want(&self) -> PubsubWant {
        PubsubWant::new(self.key.clone())
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PubsubWant {
    pub key: PubsubContentKey,
}

impl PubsubWant {
    #[must_use]
    pub fn new(key: PubsubContentKey) -> Self {
        Self { key }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PubsubFrame {
    pub key: PubsubContentKey,
    pub payload: Vec<u8>,
    pub hop_limit: u8,
}

impl PubsubFrame {
    #[must_use]
    pub fn new(key: PubsubContentKey, payload: impl Into<Vec<u8>>, hop_limit: u8) -> Self {
        Self {
            key,
            payload: payload.into(),
            hop_limit,
        }
    }

    #[must_use]
    pub fn inventory(&self) -> PubsubInventory {
        PubsubInventory::new(
            self.key.clone(),
            u64::try_from(self.payload.len()).unwrap_or(u64::MAX),
            self.hop_limit,
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum InvWantMessage {
    Inventory(PubsubInventory),
    Want(PubsubWant),
    Frame(PubsubFrame),
}

impl InvWantMessage {
    #[must_use]
    pub fn key(&self) -> &PubsubContentKey {
        match self {
            Self::Inventory(inventory) => &inventory.key,
            Self::Want(want) => &want.key,
            Self::Frame(frame) => &frame.key,
        }
    }

    #[must_use]
    pub fn stream_id(&self) -> &PubsubStreamId {
        match self {
            Self::Inventory(inventory) => &inventory.key.stream_id,
            Self::Want(want) => &want.key.stream_id,
            Self::Frame(frame) => &frame.key.stream_id,
        }
    }
}

#[async_trait]
pub trait InvWantBus: Send + Sync {
    async fn announce_inventory(
        &self,
        inventory: PubsubInventory,
        source: EventSource,
    ) -> Result<PublishReport>;

    async fn request_want(&self, want: PubsubWant, source: EventSource) -> Result<()>;

    async fn publish_frame(&self, frame: PubsubFrame, source: EventSource)
    -> Result<PublishReport>;
}
